sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -543,14 +544,57 @@ RT::PiProgram ProgramManager::getBuiltPIProgram(
 
   DeviceImplPtr Dev =
       (MustBuildOnSubdevice == PI_TRUE) ? DeviceImpl : RootDevImpl;
-  auto BuildF = [this, &M, &KSId, &ContextImpl, &Dev, Prg, &CompileOpts,
-                 &LinkOpts, &JITCompilationIsRequired, SpecConsts] {
-    auto Context = createSyclObjFromImpl<context>(ContextImpl);
-    auto Device = createSyclObjFromImpl<device>(Dev);
+  auto Context = createSyclObjFromImpl<context>(ContextImpl);
+  auto Device = createSyclObjFromImpl<device>(Dev);
+  const RTDeviceBinaryImage &Img =
+      getDeviceImage(M, KSId, Context, Device, JITCompilationIsRequired);
+
+  // Check that device supports all aspects used by the kernel
+  const RTDeviceBinaryImage::PropertyRange &ARange =
+      Img.getDeviceRequirements();
+
+#define __SYCL_ASPECT(ASPECT, ID)                                              \
+  case aspect::ASPECT:                                                         \
+    return #ASPECT;
+#define __SYCL_ASPECT_DEPRECATED(ASPECT, ID, MESSAGE) __SYCL_ASPECT(ASPECT, ID)
+// We don't need "case aspect::usm_allocator" here because it will duplicate
+// "case aspect::usm_system_allocations", therefore leave this macro empty
+#define __SYCL_SPECIAL_ASPECT_DEPRECATED(ASPECT, ID, MESSAGE)
+  auto getAspectNameStr = [](aspect AspectNum) -> std::string {
+    switch (AspectNum) {
+#include <sycl/info/aspects.def>
+#include <sycl/info/aspects_deprecated.def>
+    default:
+      throw sycl::exception(
+          errc::kernel_not_supported,
+          "Unknown aspect " + std::to_string(static_cast<unsigned>(AspectNum)));
+    }
+  };
+#undef __SYCL_SPECIAL_ASPECT_DEPRECATED
+#undef __SYCL_ASPECT_DEPRECATED
+#undef __SYCL_ASPECT
 
-    const RTDeviceBinaryImage &Img =
-        getDeviceImage(M, KSId, Context, Device, JITCompilationIsRequired);
+  for (RTDeviceBinaryImage::PropertyRange::ConstIterator It : ARange) {
+    auto KName = std::string_view((*It)->Name);
+    if (KName != "aspects")
+      continue;
```

Comment:
nit: Not used to the `std::string_view` best practices yet, but I'd try to write like this:

```suggestion
    if (It->Name != "aspects"sv)
      continue;
```

I'm also not quite sure about `It->Name` vs `(*It)->Name`, but I vaguely recall `operator->` behavior would enable that. Might be wrong though.